sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -466,6 +466,10 @@ trait CheckAnalysis extends PredicateHelper {
                         s"${field.dataType.simpleString} cannot be cast to " +
                         s"${update.newDataType.simpleString}")
                 }
+                if (!update.isNullable && field.nullable) {
+                  throw new AnalysisException(
+                    s"Cannot change optional column to required: ${field.name}")
```

Comment:
```sql
postgres=# create table a(a int null);
CREATE TABLE
postgres=# create table bb(b int not null);
CREATE TABLE
postgres=# alter table a alter column a set not null;
ALTER TABLE
postgres=# alter table bb alter column b drop not null;
ALTER TABLE
postgres=# alter table a alter column a drop not null;
ALTER TABLE
postgres=# insert into a values(null);
INSERT 0 1
postgres=# alter table a alter column a set not null;
ERROR:  column "a" contains null values
```
Postgres seems not to fail with analysis

Reply:
It's a big change if we want to allow unsafe type change (e.g. need data validation). I'd like to keep the current behavior for now.